### ISSUE TITLE
Difference in comparison of distclus KMeans with sklearn KMeans on reference dataset

### ISSUE THREAD

**Issue by EvgeniiMunin** [wearelumenai/distclus4py#8]
Hello,

I have tried to test the distclus library for static learning on Mall_Customers Kaggle dataset (https://www.kaggle.com/shwetabh123/mall-customers) and compare the clustering KMeans with benchmark sklearn KMeans method. 

The both methods were initialized with kmeans++ method, same number of iterations = 300 and the same nb_clusters = 5 but for distclus KMeans (upper image) I cannot get the same result in terms of clusters and centroids positions as for sklearn KMeans (bottom image).

`algorithm = distclus.KMeans(k=5, init='kmeans_pp', nb_iter=300)`
`algorithm = KMeans(n_clusters=5, init='k-means++', max_iter=300)`

Could I ask you to help me to understand the reason why I obtain such incoherence between the result of methods

![kmeans_distclus](https://user-images.githubusercontent.com/35193823/82742892-b6a7e100-9d63-11ea-96fa-e7f3d286c5b1.png)
![sklearn_kmeans](https://user-images.githubusercontent.com/35193823/82742895-b90a3b00-9d63-11ea-830b-29293f4051f5.png)

**Comment by GENETAY** [comment]
Hello EvgeniiMinin, thank you for your comment.

I had little time to reproduce your simulations today but I did it I think for the sklearn.cluster.KMeans algorithm. I will keep you informed as soon as I have news about our algorithm.
I have got a question for you so far:
Have you runned sklearn.cluster.KMeans on the whole dataset (5 features) with 0 and 1 for Male and Female or have you runned it on a part of it, namely on the sub-dataset made of the two last features of Mall_Customers named 'annual income (k$)' and 'Spending-Score (1-100)'?

I am looking foward to reading from you.

**Comment by EvgeniiMunin** [comment]
Hello Edouard, thank you for your response !

Yes I've taken 2 features "Annual Income (k$)" and "Spending Score (1-100)" then renamed them for more convenience and then fitted KMeans on the whole dataset of 200 samples

`df.rename(columns={"Annual Income (k$)": 'income', 'Spending Score (1-100)': 'score'}, inplace=True)
X2 = df[['income', 'score']].iloc[:, :].values
algorithm.fit(X2)
labels2 = algorithm.labels_
centroids2 = algorithm.cluster_centers_
`
First my doubt was to fix the random seed (to fix the way of centroids initialization) but I suppose this was not the reason

**Comment by GENETAY** [comment]
Ok, that is good to know.
I have asked you this question because something strikes me : 
If you take a look at the second picture, the one  corresponding to the sklearn.KMeans,  the colored polygon behind the data points are called voronoi cells and are theoretically such that all the data points of a given cluster lie inside its corresponding cells : green points lie inside the orange polygon, yellow points lie inside the grey polygon, et cetera...

But I find it weird that it is not the case for the 1st picture, the one corresponding to distclus.KMeans. It must be so. We may have made a mistake in our code but before any futher investigation, I would like to check something:
**it looks like distclus.KMeans was first run on the whole data set (5 features) and then the print of this clustering made in 5 dimensions was plotted on the 2 dimensional plan [income,score] only. If it is the case, both KMeans were not run on the same dataset. Are you sure you ran the distclus.KMeans on the exact same 2 dimensional dataset as for sklearn.KMeans?**
I had no time today to work on it but be sure I will! Best regards.

**Comment by GENETAY** [comment]
Thanks you for your remark. I managed to reproduce your results. It is indeed weird. It is very likely that we have made an error somewhere. I will search for it! It must be corrected.

**Comment by EvgeniiMunin** [comment]
Concerning the point that KMeans was probably fitted on 5 dimensions tensor and then the centroids were predicted on 2 dimensions, I think that in this case we would have a problem of dimensions incoherence. I think that distclus model is robust to such kind of error

**Comment by b3j0f** [comment]
I added a bugfix in order to reduce differences between both kmeans implementation. @EvgeniiMunin could you try the last version and give me a feedback ?

**Comment by EvgeniiMunin** [comment]
Yes, sure, I will try the updated version

**Comment by EvgeniiMunin** [comment]
I'm currently trying to update the library from current repo. How can do it properly ? Do I need to rebuild it from Makefile and create another python environment or there is another way to do it ?

**Comment by EvgeniiMunin** [comment]
Hello, sorry for some delay, I have checked the updated implementation of the algorithm and now the clustering is working correctly ! Thank you  @b3j0f, @GENETAY for your time on fixing this problem.

And could you precise what was the changes applied to make the algorithm work ?

![image](https://user-images.githubusercontent.com/35193823/84442476-fadd2180-ac3d-11ea-9196-239ed39ab872.png)

**Comment by b3j0f** [comment]
Thanks for your feedback.

According to our [cffi](https://cffi.readthedocs.io/en/latest/) implementation (python -> go binding), we considered input data such as [c continguous data](https://stackoverflow.com/a/26999092/13116126).

Therefore, we forced the memory transformation if it was not the case.

And I added generic python collection such as new input data type...

Can I ask you how you discovered our library ? And what do you want to do with it ?

**Comment by EvgeniiMunin** [comment]
I've been in contact with your colleagues M. Christian Dufrechou, M. Paul Gay et M. Chemsdin Naamane for the question of implementation of distclus library in my task of detection of errors in GNSS signals.

And I've first tried the library on some reference data before passing to the GNSS encoded data.
